### DISCUSSION TITLE
pudb v2022.1.2 breaks automatically on exception:

### DISCUSSION THREAD

**Discussion by shaoran** [inducer/pudb#560] [Q&A]
Hi

consider this code with an obvious exception:

```python
import asyncio
import sys 
import pudb

async def has_an_error():
    print("This function has en error")
    a = {}

    b = a["b"] + 1 

    return b


async def main():
    pudb.set_trace()

    try:
        n = await has_an_error()
    except asyncio.CancelledError:
        raise
    except:
        print("Some error")

    print("n =", n)

if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
```

I press *n* once to advance the code to `await has_an_error()` and if I press *n* again, then pudb breaks automatically and jumps into the function that raised the error. This is a behaviour that did not occur with older pudb versions like 2022.1.1, as instead of breaking and jumping into the failing line, it would only show me that an exception was raised (and I would press *e* to inspect t he exception) and I would remain where I was. This is bothersome because I sometimes **know** that the call has an exception and I want to debug the exception handler, so I don't to jump into the function that raises the error and specially with asyncio, depending how you started the task, you might not be able to come back to the same place again without restarting everything.

I don't see any obvious option on the configuration (when pressing ctrl-p). How can I get the old behaviour?

**Comment by asmeurer** [comment]
I'm not sure because I haven't followed every change here, but this sounds like a change to the underlying standard library `bdb` debugger which pudb uses. Check if you're using the same version of Python that you were using before.

**Reply by shaoran** [reply]
> I'm not sure because I haven't followed every change here, but this sounds like a change to the underlying standard library `bdb` debugger which pudb uses. Check if you're using the same version of Python that you were using before.

It's the same python version:

```console
$ pip install pudb==2022.1.2
$ python case2.py
```
when when I press *n*, it jumps into `has_an_error()`.

On the same terminal, no change of environment variables
```
$ pip install pudb==2022.1.1
$ python case2.py
```

then it doesn't jump into `has_an_error()`.

Note that after doing `pip install`, no other package gets updated, so the same python version, same everything. So something in the `2022.1.2` released changed so that when an exception is raised, the ui jumps into the line that raises the exception. The question is, can anybody confirm this behaviour with my example script?

**Reply by inducer** [reply]
https://github.com/inducer/pudb/compare/v2022.1.1...v2022.1.2

I don't see anything in these changes that might have a causal connection to the behavior you describe. Could you try and bisect?

**Reply by shaoran** [reply]
OK, I will, but it will take a while though.

**Reply by shaoran** [reply]
OK, I did the bisect and this is the commit that bisect found:

```
98ccafc49c1d3b31a74987eebb6cb11ba2c89998 is the first bad commit
commit 98ccafc49c1d3b31a74987eebb6cb11ba2c89998
Author: Andreas Kloeckner <inform@tiker.net>
Date:   Thu May 19 16:34:04 2022 -0500

    get_shortened_stack: Use either frame *or* backtrace to produce stack

 pudb/debugger.py | 2 ++
 1 file changed, 2 insertions(+)
```

Should I create a bug report?

**Reply by inducer** [reply]
Yes, please! Thanks.
